source/fab/engine.py

```diff
@@ -0,0 +1,190 @@
+            new_artifact = None
+            for pathmap in self._pathmaps:
+                if pathmap.match(artifact.location):
+                    new_artifact = Artifact(artifact.location,
+                                            pathmap.filetype,
+                                            pathmap.state)
+            # Assuming we found a match and were able
+            # to create the artifact, return it so that
+            # it can be added to the queue
+            if new_artifact is not None:
+                # Also store its hash in the file database
+                file_info = FileInfoDatabase(self._database)
+                file_info.add_file_info(artifact.location,
+                                        new_artifact.hash)
+                new_artifacts.append(new_artifact)
+
+        elif artifact.state is Analysed:
+
+            # Work out whether this artifact needs to be
+            # included in the build or not - tag it as having
+            # been seen so the system knows it is Analysed
+            required = False
+            for definition in artifact.defines:
+                if definition in discovery:
+                    new_discovery[definition] = "Seen"
+                    required = True
+
+            # Assuming it is needed, check its
+            # dependencies to know what needs doing
+            if required:
+                compiled = [False]*len(artifact.depends_on)
+                for idep, dependency in enumerate(artifact.depends_on):
+                    if dependency in discovery:
+                        # Are the dependencies compiled?
+                        if discovery[dependency] == "Compiled":
+                            compiled[idep] = True
+                    else:
+                        # If the dependency isn't in the list at all yet
+                        # then add an entry so the system knows we are
+                        # expecting it later (for the above check)
+                        new_discovery[dependency] = "HeardOf"
+
+                # If the dependencies are satisfied (or there weren't
+                # any) then the file can be compiled now
+                if len(compiled) == 0 or all(compiled):
+                    for definition in artifact.defines:
+                        task = self._taskmap[(artifact.filetype,
+                                              artifact.state)]
+                        new_artifacts.extend(task.run([artifact]))
+                        new_discovery[definition] = "Compiled"
+                else:
+                    # If the dependencies weren't all satisfied then
+                    # back on the queue for another pass later
+                    new_artifacts.append(artifact)
+            else:
+                # If it wasn't required it could be later, so
+                # put it back on the queue, unless the target
+                # has been compiled, in which case it wasn't
+                # needed at all!
```

Comment:
The comment talks about an error condition but the code doesn't seem to handle it. Maybe a ToDo?

Reply:
I'm not sure it is an error condition.  It's the case where some of the Analysed files might not be needed to compile the target (this comes up in the preprocessing source test; depending on the macro you only need one of "stay" or "go now").  So this is just cycling the Analysed artifacts back onto the queue until they are either handled above (because their dependencies become Compiled) or the target gets compiled without them (so they weren't needed, and are dropped from the queue to let it finish)